File: backend/app/workflow.py
"""Single source of truth for the Wet Lab -> Bioinfo data approval pipeline.

Maps 1:1 onto the 12-step flow diagram. Steps 5 and 9 of that diagram ("prepares
consolidated Excel/PDF") are work done outside the dashboard, so they are not separate
dashboard actions — preparing and uploading are one action here. `step` is the label
from the diagram; `id` is the internal stage key.

  diagram step        stage id               who
  1  raw data -> drive    raw_data_placed        Wet Lab
  2  dashboard updated    dashboard_updated      Wet Lab
  3  notification         notified               Primary Team (automatic)
  4  raw CSV + Excel      drive_checked          Primary Team
  5-6 consolidated Excel  excel_uploaded         Primary Team
  7  head approves        primary_head_review    Primary Team Head
  7  QC approves          wetlab_qc_review       Wet Lab
  8  bioinfo sign-off     bioinfo_review         Bioinfo Team Head
  9-10 consolidated PDF   pdf_uploaded           Primary Team
  11 PDF approval         wetlab_final_review    Wet Lab
  12 status + notes       completed              —
"""

ROLES = ["wetlab", "primary_team", "primary_head", "bioinfo_head"]

STAGES = [
    {"id": "raw_data_placed", "step": "1", "role": "wetlab",
     "title": "Raw data on common drive",
     "desc": "FastQ, CSV, Excel and Md5sum files placed on the common drive."},
    {"id": "dashboard_updated", "step": "2", "role": "wetlab",
     "title": "Marked as transferred",
     "desc": "Wet Lab marked the run as transferred so the Primary Team is notified."},
    {"id": "notified", "step": "3", "role": "primary_team",
     "title": "Primary Team notified",
     "desc": "Alert raised in the dashboard that new run data is ready on the common drive."},
    {"id": "drive_checked", "step": "4", "role": "primary_team",
     "title": "Raw CSV & Excel upload",
     "desc": "Raw data checked on the common drive; CSV and Excel uploaded to the dashboard as backup."},
    {"id": "excel_uploaded", "step": "5-6", "role": "primary_team",
     "title": "Consolidated Excel upload",
     "desc": "Consolidated Excel (raw data plus essential details and analysis) uploaded for review and approvals."},
    {"id": "primary_head_review", "step": "7", "role": "primary_head",
     "title": "Primary Team Head approval",
     "desc": "First approval on the consolidated Excel. Flag errors directly in the sheet; a note is optional."},
    {"id": "wetlab_qc_review", "step": "7", "role": "wetlab",
     "title": "Wet Lab QC approval",
     "desc": "Second of the two-step approval — Wet Lab checks QC and approves the consolidated Excel."},
    {"id": "bioinfo_review", "step": "8", "role": "bioinfo_head",
     "title": "Bioinfo Team Head approval",
     "desc": "Final sign-off on data quality for the consolidated Excel."},
    {"id": "pdf_uploaded", "step": "9-10", "role": "primary_team",
     "title": "Consolidated PDF upload",
     "desc": "Consolidated PDF (all details from the Excel in report format) uploaded for final review."},
    {"id": "wetlab_final_review", "step": "11", "role": "wetlab",
     "title": "Wet Lab PDF approval",
     "desc": "Wet Lab views the consolidated PDF and approves, with optional notes or corrections."},
    {"id": "completed", "step": "12", "role": None,
     "title": "Samples approved, moved to Tertiary team for analysis",
     "desc": "Every approval is in; the samples are handed off to the Tertiary team for analysis. "
             "All approvals, timestamps and Wet Lab notes remain available to the Primary Team."},
]

STAGE_BY_ID = {s["id"]: s for s in STAGES}
STAGE_INDEX = {s["id"]: i for i, s in enumerate(STAGES)}
FIRST_STAGE_ID = STAGES[0]["id"]

REVIEW_STAGE_IDS = {"primary_head_review", "wetlab_qc_review", "bioinfo_review", "wetlab_final_review"}

# A reviewer requesting changes sends the run back to the upload it was reviewing.
REJECT_TARGET = {
    "primary_head_review": "excel_uploaded",
    "wetlab_qc_review": "excel_uploaded",
    "bioinfo_review": "excel_uploaded",
    "wetlab_final_review": "pdf_uploaded",
}

FILE_REQUIRED_STAGES = {"drive_checked", "excel_uploaded", "pdf_uploaded"}

# The Wet Lab login covers both the initial transfer (logged as the generic "Wet Lab
# Team") and these two approval steps, which are actually done by a named reviewer —
# so the history for these specific stages is attributed to them by name instead of
# the account's display name.
ACTOR_NAME_OVERRIDE = {
    "wetlab_qc_review": "Dr. Sivasankar",
    "wetlab_final_review": "Dr. Sivasankar",
}


def actor_name_for(stage_id: str, account_full_name: str) -> str:
    return ACTOR_NAME_OVERRIDE.get(stage_id, account_full_name)


# Steps that are another team's internal mechanics, hidden from roles with no part in
# them — the raw CSV/Excel backup and the automatic "Primary Team notified" checkpoint
# are Primary Team's own business, so Wet Lab and both team heads should never see
# that step, its files, or its history, in the API response itself (not just the UI).
HIDDEN_STAGES_FOR_ROLE = {
    "wetlab": {"notified", "drive_checked"},
    "primary_head": {"notified", "drive_checked"},
    "bioinfo_head": {"notified", "drive_checked"},
    "primary_team": set(),
}


def can_see_stage(stage_id: str, role: str) -> bool:
    return stage_id not in HIDDEN_STAGES_FOR_ROLE.get(role, set())


def visible_history(history, role):
    return [h for h in history if can_see_stage(h.stage_id, role)]


def stage_role(stage_id: str):
    return STAGE_BY_ID[stage_id]["role"]


def next_stage_after(stage_id: str, action: str) -> str:
    """The stage a run moves to, given where it is and the action just taken."""
    if stage_id in REVIEW_STAGE_IDS and action == "rejected":
        return REJECT_TARGET[stage_id]
    return STAGES[STAGE_INDEX[stage_id] + 1]["id"]

File: frontend/src/workflow.js
// Mirrors backend/app/workflow.py — keep the two in sync.
//
// Wording rule: nothing here mentions step numbers. `title` reads as what the run is
// waiting on ("Consolidated Excel upload"), `heading` is the imperative for whoever
// owns it ("Upload the consolidated Excel"), `action` is the button label, `logVerb`
// is the past-tense phrase used in the Status timeline ("Sethu uploaded the
// consolidated Excel") so a plain completion reads like a sentence, the same way an
// approval reads "Tamilarasu approved".

// All four roles sit in the orange/blue theme: the two teams take the bright tones,
// the two heads the deep tones, which keeps them on-brand and still tells them apart.
// `color`/`soft` are for light surfaces; `onDark` is the variant with enough contrast
// for the dark header bar.
export const ROLES = {
  wetlab: { id: "wetlab", label: "Wet Lab", color: "#c1440e", soft: "#fdeee6", onDark: "#f4a882" },
  primary_team: { id: "primary_team", label: "Primary Team", color: "#1668c1", soft: "#e7f1fd", onDark: "#8dc0f2" },
  primary_head: { id: "primary_head", label: "Primary Team Head", color: "#0b3462", soft: "#e2eaf4", onDark: "#adc8e6" },
  bioinfo_head: { id: "bioinfo_head", label: "Bioinfo Team Head", color: "#8a3a12", soft: "#f7ebe3", onDark: "#dfa77f" },
};

export const STAGES = [
  {
    id: "raw_data_placed",
    role: "wetlab",
    title: "Raw data on common drive",
    logVerb: "placed the raw data on the common drive",
    desc: "FastQ, CSV, Excel and Md5sum files placed on the common drive.",
  },
  {
    id: "dashboard_updated",
    role: "wetlab",
    title: "Marked as transferred",
    logVerb: "marked the run as transferred",
    desc: "Wet Lab confirmed the transfer, which notifies the Primary Team.",
  },
  {
    id: "notified",
    role: "primary_team",
    title: "Primary Team notified",
    desc: "Alert raised that new run data is ready on the common drive.",
  },
  {
    id: "drive_checked",
    role: "primary_team",
    title: "Raw CSV & Excel upload",
    logVerb: "uploaded the raw CSV & Excel",
    heading: "Upload raw CSV & Excel",
    desc: "Check the raw data on the common drive, then upload whichever of the CSV and Excel you have here as backup. One of them is enough, both is fine too.",
    action: "Upload raw backup",
    needsFiles: true,
  },
  {
    id: "excel_uploaded",
    role: "primary_team",
    title: "Consolidated Excel upload",
    logVerb: "uploaded the consolidated Excel",
    heading: "Upload consolidated Excel",
    desc: "Upload the consolidated Excel (raw data plus the essential details and analysis) to send it for approval.",
    action: "Upload & send for approval",
    needsFiles: true,
  },
  {
    id: "primary_head_review",
    role: "primary_head",
    title: "Primary Team Head approval",
    heading: "Approve the consolidated Excel",
    desc: "Check the consolidated Excel and approve it, or flag errors in the sheet and send it back. A note is optional.",
    review: true,
  },
  {
    id: "wetlab_qc_review",
    role: "wetlab",
    title: "Wet Lab QC approval",
    heading: "Check QC and approve the Excel",
    desc: "Check the QC and approve the consolidated Excel, or send it back with a note.",
    review: true,
  },
  {
    id: "bioinfo_review",
    role: "bioinfo_head",
    title: "Bioinfo Team Head approval",
    heading: "Final sign-off on the Excel",
    desc: "Final sign-off on data quality for the consolidated Excel.",
    review: true,
  },
  {
    id: "pdf_uploaded",
    role: "primary_team",
    title: "Consolidated PDF upload",
    logVerb: "uploaded the consolidated PDF",
    heading: "Upload consolidated PDF",
    desc: "Upload the consolidated PDF (all the details from the Excel in report format) for final review.",
    action: "Upload PDF",
    needsFiles: true,
  },
  {
    id: "wetlab_final_review",
    role: "wetlab",
    title: "Wet Lab PDF approval",
    heading: "Approve the report PDF",
    desc: "View the consolidated PDF and approve it, adding notes or corrections if needed.",
    review: true,
  },
  {
    id: "completed",
    role: null,
    title: "Samples approved, moved to Tertiary team for analysis",
    desc: "Every approval is in; the samples are handed off to the Tertiary team for analysis. All timestamps and notes are recorded against the run.",
  },
];

export const STAGE_BY_ID = Object.fromEntries(STAGES.map((s) => [s.id, s]));
export const STAGE_INDEX = Object.fromEntries(STAGES.map((s, i) => [s.id, i]));

/** File inputs each upload step asks for. */
export const UPLOAD_SLOTS = {
  drive_checked: [
    { key: "csv", label: "Raw CSV file", accept: ".csv,text/csv" },
    { key: "excel", label: "Raw Excel file", accept: ".xlsx,.xls" },
  ],
  excel_uploaded: [{ key: "excel", label: "Consolidated Excel", accept: ".xlsx,.xls" }],
  pdf_uploaded: [{ key: "pdf", label: "Consolidated PDF", accept: ".pdf" }],
};

/** Steps whose slots aren't all mandatory — just at least one of them. The raw
 * backup is sometimes only a CSV, sometimes only an Excel, sometimes both. */
export const UPLOAD_ANY_OF = new Set(["drive_checked"]);

/** The upload steps the Primary Team owns, in the order they have to happen. */
export const PRIMARY_TEAM_UPLOADS = ["drive_checked", "excel_uploaded", "pdf_uploaded"];

/**
 * Steps that are another team's internal mechanics, hidden from roles with no part
 * in them — the raw CSV/Excel backup and the Primary Team's own alert are Primary
 * Team business, so Wet Lab and the heads never see those steps or their files.
 */
const HIDDEN_STAGES = {
  wetlab: ["notified", "drive_checked"],
  primary_head: ["notified", "drive_checked"],
  bioinfo_head: ["notified", "drive_checked"],
  primary_team: [],
};

export function canSeeStage(stageId, role) {
  return !(HIDDEN_STAGES[role] || []).includes(stageId);
}

export function visibleStages(role) {
  return STAGES.filter((s) => canSeeStage(s.id, role));
}

/**
 * Every role can see the current workflow position, even where files and detailed
 * history for another team's internal step remain restricted.
 */
export function stageLabelFor(run, role) {
  const stage = STAGE_BY_ID[run.current_stage];
  if (!stage) return "";
  return stage.title;
}

export function stageOf(run) {
  return STAGE_BY_ID[run.current_stage];
}

export function isAwaiting(run, role) {
  const stage = STAGE_BY_ID[run.current_stage];
  return run.status === "in_progress" && stage && stage.role === role;
}

export function wasSentBack(run) {
  const last = run.history?.[run.history.length - 1];
  return last?.action === "rejected" || last?.action === "reset";
}

/**
 * One label for where a run stands, including whether a send-back has been
 * re-uploaded yet. Fields come from the runs endpoints. Blue reads as "moving
 * normally" (whether that's still in progress or already done), and only the one
 * state that actually needs a person to act — a send-back awaiting re-upload —
 * gets orange, so that color keeps its meaning instead of covering every row.
 */
export function runState(run) {
  if (run.status === "completed") {
    return { key: "completed", label: "Completed", chip: "bg-brand-100 text-brand-800" };
  }
  if (run.awaiting_reupload) {
    return {
      key: "needs_reupload",
      label: "Needs re-upload",
      chip: "bg-accent-100 text-accent-800",
      urgent: true,
    };
  }
  if (run.rejection_count > 0) {
    return { key: "reuploaded", label: "Back in review", chip: "bg-brand-50 text-brand-700" };
  }
  return { key: "in_progress", label: "In progress", chip: "bg-brand-50 text-brand-700" };
}
